zerver/lib/user_groups.py

```diff
@@ -219,17 +226,96 @@ def access_user_group_for_setting(
 
     if (
         permission_configuration.allowed_system_groups
-        and user_group.name not in permission_configuration.allowed_system_groups
+        and setting_group.name not in permission_configuration.allowed_system_groups
     ):
         raise JsonableError(
             _("'{setting_name}' setting cannot be set to '{group_name}' group.").format(
-                setting_name=setting_name, group_name=user_group.name
+                setting_name=setting_name, group_name=setting_group.name
             )
         )
 
+
+def update_or_create_user_group_for_setting(
+    realm: Realm,
+    direct_members: List[int],
+    direct_subgroups: List[int],
+    current_setting_value: Optional[UserGroup],
+) -> UserGroup:  # nocoverage
+    if current_setting_value is not None and not hasattr(current_setting_value, "named_user_group"):
```

Comment:
We can probably avoid one query here by updating the UserGroup query to have `select_related(can_mention_group__named_user_group)`.

Reply:
That might be worth doing -- but seems reasonable to leave for a follow-up.